kyo-core/shared/src/main/scala/kyo/chunks.scala

```diff
@@ -0,0 +1,467 @@
+package kyo
+
+import Chunks.Indexed
+import scala.annotation.tailrec
+import scala.reflect.ClassTag
+
+sealed abstract class Chunk[T] derives CanEqual:
+
+    import Chunks.internal.*
+
+    //////////////////
+    // O(1) methods //
+    //////////////////
+
+    def size: Int
+
+    final def isEmpty: Boolean = size == 0
+
+    final def take(n: Int): Chunk[T] =
+        dropLeftAndRight(0, size - Math.min(Math.max(0, n), size))
+
+    final def dropLeft(n: Int): Chunk[T] =
+        dropLeftAndRight(Math.min(size, Math.max(0, n)), 0)
+
+    final def dropRight(n: Int): Chunk[T] =
+        dropLeftAndRight(0, Math.min(size, Math.max(0, n)))
+
+    final def slice(from: Int, until: Int): Chunk[T] =
+        dropLeftAndRight(Math.max(0, from), size - Math.min(size, until))
+
+    final def dropLeftAndRight(left: Int, right: Int): Chunk[T] =
+        @tailrec def loop(c: Chunk[T], left: Int, right: Int): Chunk[T] =
+            val size = c.size - left - right
+            if size <= 0 then Chunks.init
+            else
+                c match
+                    case Drop(chunk, dropLeft, dropRight, _) =>
+                        Drop(chunk, left + dropLeft, right + dropRight, size)
+                    case Append(chunk, value, size) if right > 0 =>
+                        loop(chunk, left, right - 1)
+                    case _ =>
+                        Drop(c, left, right, size)
+            end if
+        end loop
+        loop(this, left, right)
+    end dropLeftAndRight
+
+    final def append(v: T): Chunk[T] =
+        Append(this, v, size + 1)
+
+    final def last: T =
+        @tailrec def loop(c: Chunk[T], index: Int): T =
+            c match
+                case c if index >= c.size || index <= 0 =>
+                    throw new NoSuchElementException
+                case c: Append[T] =>
+                    if index == c.size - 1 then
+                        c.value
+                    else
+                        loop(c.chunk, index)
+                case c: Drop[T] =>
+                    loop(c.chunk, index + c.dropLeft)
+                case c: Indexed[T] =>
+                    c(index)
+        loop(this, this.size - 1)
+    end last
+
+    //////////////////
+    // O(n) methods //
+    //////////////////
+
+    final def concat(other: Chunk[T]): Chunk[T] =
+        if isEmpty then other
+        else if other.isEmpty then this
+        else
+            val s     = size
+            val array = new Array[T](s + other.size)(using ClassTag(classOf[Any]))
+            this.copyTo(array, 0)
+            other.copyTo(array, s)
+            Compact(array)
+        end if
+    end concat
+
+    final def map[U: Flat, S](f: T => U < S): Chunk[U] < S =
+        if isEmpty then Chunks.init
+        else
+            val size    = this.size
+            val indexed = this.toIndexed
+            Loops.indexed(Chunks.init[U]) { (idx, acc) =>
```

Comment:
Shouldn't it be cheaper to pre-allocate a backing Array instead of growing the `init`? We can still use the `Loops` to fill it...

Reply:
It's not ideal because it'd introduce mutable state in continuations, similarly to ZIO's approach. If an effect handler calls a continuation multiple times, the continuation could have a different behavior on each call.